File: java/core/src/main/java/de/zannagh/eunomia/clients/WsFrame.java
package de.zannagh.eunomia.clients;

import com.google.gson.JsonElement;

/**
 * A tagged frame the relay pushes over the WebSocket. {@link #type} is {@code "envelope"} (a single relayed
 * {@link PacketEnvelope} update) or {@code "store_sync"} (a batch {@code StoreSyncPayload} snapshot); {@link #data}
 * is the corresponding object. The tag lets the client route the frame without guessing from its shape.
 */
public class WsFrame {

    public String type;

    public JsonElement data;

    public WsFrame() {
    }
}

File: csharp/src/Eunomia.Server.Core/Communication/WsFrame.cs
// Copyright (c) 2026, zannagh. All rights reserved.
// See License in the project root for license information.

namespace Eunomia.Server.Core.Communication;

/// <summary>
/// Thin tagged wrapper for every server-&gt;client push, so the Java client can distinguish a
/// per-update <c>PacketEnvelope</c> from a batch <c>StoreSyncPayload</c> without a second frame
/// type: <c>{ "type": "envelope", "data": &lt;PacketEnvelope&gt; }</c> or
/// <c>{ "type": "store_sync", "data": &lt;StoreSyncPayload&gt; }</c>.
/// </summary>
/// <typeparam name="T">The payload type carried in <c>data</c>.</typeparam>
public sealed record WsFrame<T>(string Type, T Data)
{
    public static WsFrame<T> Envelope(T data) => new("envelope", data);

    public static WsFrame<T> StoreSync(T data) => new("store_sync", data);
}
